### ISSUE TITLE
begin/end not indented correctly

### ISSUE THREAD

**Issue by fdw** [sqlfluff/sqlfluff#2360]
## Expected Behaviour
Blocks with `begin` and `end` should be indented exactly `tab_space_size` more than their surrounding block, f.e.

```
[sqlfluff:rules]
tab_space_size = 2
```

```sql
begin
  begin try
    declare @message nvarchar(12);
  end try
end
```

## Observed Behaviour
The given SQL (with wrong indentation) is not changed:
```sql
begin
    begin try
        declare @message nvarchar(12);
    end try
end
```

## Steps to Reproduce
1. Have file with `begin`/`end` and incorrect indentation
2. Define `tab_space_size` to a different value.
3. Run `sqlfluff fix`

## Dialect
T-SQL

## Version
`sqlfluff, version 0.9.1` (from git, ~2 days ago)
`Python 3.10.1`

## Configuration
```
[sqlfluff]
dialect = tsql

[sqlfluff:rules]
tab_space_size = 2
indent_unit = space
comma_style = trailing

[sqlfluff:indentation]
indented_joins = True

[sqlfluff:rules:L010] # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L011] # Aliasing
aliasing = implicit

[sqlfluff:rules:L012] # Aliasing
aliasing = explicit

[sqlfluff:rules:L013] # Aliasing
allow_scalar = True

[sqlfluff:rules:L014] # Unquoted identifiers
extended_capitalisation_policy = pascal

[sqlfluff:rules:L040] # Capitalization of NULL
capitalisation_policy: upper
```

**Comment by tunetheweb** [comment]
Looks to me like `TRY` is not supported yet.

```
% sqlfluff parse --dialect tsql test.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    unparsable:                                               !! Expected: 'FileSegment'
[L:  1, P:  1]      |        raw:                                                  'begin'
[L:  1, P:  6]      |        newline:                                              '\n'
[L:  2, P:  1]      |        whitespace:                                           '    '
[L:  2, P:  5]      |        raw:                                                  'begin'
[L:  2, P: 10]      |        whitespace:                                           ' '
[L:  2, P: 11]      |        raw:                                                  'try'
[L:  2, P: 14]      |        newline:                                              '\n'
[L:  3, P:  1]      |        whitespace:                                           '        '
[L:  3, P:  9]      |        raw:                                                  'declare'
[L:  3, P: 16]      |        whitespace:                                           ' '
[L:  3, P: 17]      |        raw:                                                  '@message'
[L:  3, P: 25]      |        whitespace:                                           ' '
[L:  3, P: 26]      |        raw:                                                  'nvarchar'
[L:  3, P: 34]      |        raw:                                                  '('
[L:  3, P: 35]      |        raw:                                                  '12'
[L:  3, P: 37]      |        raw:                                                  ')'
[L:  3, P: 38]      |        raw:                                                  ';'
[L:  3, P: 39]      |        newline:                                              '\n'
[L:  4, P:  1]      |        whitespace:                                           '    '
[L:  4, P:  5]      |        raw:                                                  'end'
[L:  4, P:  8]      |        whitespace:                                           ' '
[L:  4, P:  9]      |        raw:                                                  'try'
[L:  4, P: 12]      |        newline:                                              '\n'
[L:  5, P:  1]      |        raw:                                                  'end'
[L:  5, P:  4]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'begin\n begin try\n
                       | declare @mes...'
```

Now, as raised in #2345 `fix` doesn't warn you of this (but probably should!)

If you try the same with this:

```sql
begin
    begin
        declare @message nvarchar(12);
    end
end
```

It correctly flags:

```
% sqlfluff lint --dialect tsql test.sql
== [test.sql] FAIL                                                                                                                                                                                                                                         
L:   2 | P:   5 | L003 | Line over-indented compared to line #1
L:   3 | P:   9 | L003 | Line over-indented compared to line #1
L:   4 | P:   5 | L003 | Line over-indented compared to line #1
All Finished 📜 🎉!
```

So do you want to update the title to "`TRY` not supported for T-SQL"?

**Comment by fdw** [comment]
You're right, this only happens with `begin try`. However, I thought it should be implemented (#1541 and #2044) :/

**Comment by tunetheweb** [comment]
The CATCH part is not optional - should it be?
Currently, without it, SQLFluff can't understand the statement.

This works:

```sql
begin
    begin try
        declare @message nvarchar(12);
    end try
    begin catch
        throw;
    end catch
end
```

```
% sqlfluff lint --dialect tsql test.sql
== [test.sql] FAIL                                                                                                                                                                                                                                         
L:   2 | P:   5 | L003 | Line over-indented compared to line #1
L:   3 | P:   9 | L003 | Line over-indented compared to line #1
L:   4 | P:   5 | L003 | Line over-indented compared to line #1
L:   5 | P:   5 | L003 | Line over-indented compared to line #1
L:   6 | P:   9 | L003 | Line over-indented compared to line #1
L:   6 | P:   9 | L010 | Keywords must be consistently lower case.
L:   7 | P:   5 | L003 | Line over-indented compared to line #1
All Finished 📜 🎉!
```

**Comment by fdw** [comment]
Again, you're right. Now I'm left wondering which part of the file cannot be parsed :/
Unfortunately, I cannot share it, and it is too long to anonymize. I've tried to run `sqlfluff parse`, but I only get
```
[L: 20, P:  1]      |    unparsable:                                               !! Expected: 'FileSegment'
[...]
L:  20 | P:   1 |  PRS | Line 20, Position 1: Found unparsable section: 'create or alter
                       | procedure <blah>'
```
Everything in between is only matched to `raw`.

`create or alter procedure` works on its own, however.

Any ideas how I can debug this?

**Comment by tunetheweb** [comment]
Unfortunately it's just a matter of deleting bits of the query to retry to get a minimal reproducible example.

Going to close this for now as seems to be a parsing issue causing it.

**Comment by fdw** [comment]
I actually found several problems with the code and have opened issues for them. Thanks for your help! 🙂
